components/raftstore/src/store/fsm/store.rs

```diff
@@ -1176,29 +1172,60 @@ impl<EK: KvEngine, ER: RaftEngine> RaftBatchSystem<EK, ER> {
             .register_admin_observer(100, BoxAdminObserver::new(SplitObserver));
 
         let workers = Workers {
-            split_check_worker,
-            region_worker: Worker::new("snapshot-worker"),
             pd_worker,
-            consistency_check_worker: Worker::new("consistency-check"),
-            cleanup_worker: Worker::new("cleanup-worker"),
-            raftlog_gc_worker: Worker::new("raft-gc-worker"),
+            background_worker,
             coprocessor_host: coprocessor_host.clone(),
         };
+        mgr.init()?;
+        let region_runner = RegionRunner::new(
+            engines.kv.clone(),
+            mgr.clone(),
+            cfg.value().snap_apply_batch_size.0 as usize,
+            cfg.value().use_delete_range,
+            workers.coprocessor_host.clone(),
+            self.router(),
+        );
+        let region_scheduler = workers
+            .background_worker
+            .start_with_timer("snapshot-worker", region_runner);
+
+        let raftlog_gc_runner = RaftlogGcRunner::new(self.router(), engines.clone());
+        let raftlog_gc_scheduler = workers
+            .background_worker
+            .start_with_timer("raft-gc-worker", raftlog_gc_runner);
+
+        let compact_runner = CompactRunner::new(engines.kv.clone());
```

Comment:
Compact task may take minutes, could it block all threads in thread pool?

Reply:
No, it will only block one thread, not all.